pkg/policy/selectorcache.go

```diff
@@ -0,0 +1,336 @@
+// callbacks whenever the set of selected numeric identities change for
+// the CachedSelectors pushed by it.
+type CachedSelectionUser interface {
+	// IdentitySelectionUpdated implementations MUST NOT call back
+	// to selector cache while executing this function!
+	IdentitySelectionUpdated(selector CachedSelector, selections, added, deleted []identity.NumericIdentity)
+}
+
+// IdentitySelector is the internal interface for all selectors in the
+// selector cache.
+//
+// IdentitySelector represents the mapping of an EndpointSelector
+// to a slice of identities. These mappings are updated via two
+// different processes:
+//
+// 1. When policy rules are changed these are added and/or deleted
+// depending on what selectors the rules contain. Cached selections of
+// new IdentitySelectors are pre-populated from the set of currently
+// known identities.
+//
+// 2. When reachacble identities appear or disappear, either via local
+// allocation (CIDRs), or via the KV-store (remote endpoints). In this
+// case all existing IdentitySelectors are walked through and their
+// cached selections are updated as necessary.
+//
+// In both of the above cases the set of existing IdentitySelectors is
+// write locked.
+//
+// To minimize the upkeep the identity selectors are shared accross
+// all IdentityPolicies, so that only one copy exists for each
+// IdentitySelector. Users of the SelectorCache take care of creating
+// IdentitySelectors as needed by identity policies. The set of
+// IdentitySelectors is read locked during an IdentityPolicy update so
+// that the the policy is always updated using a coherent set of
+// cached selections.
+//
+// IdentitySelector is used as a map key, so it must not be implemented by a
+// map, slice, or a func, or a runtime panic will be triggered. In all
+// cases below IdentitySelector is being implemented by structs.
+type IdentitySelector interface {
+	CachedSelector
+	addUser(CachedSelectionUser) (added bool)
+	removeUser(CachedSelectionUser) (last bool)
+	notifyUsers(added, deleted []identity.NumericIdentity)
+}
+
+// SelectorCache caches identities, identity selectors, and the
+// subsets of identities each selector selects.
+type SelectorCache struct {
+	mutex lock.RWMutex
+
+	// idCache contains all known identities as informed by the
+	// kv-store and the local identity facility via our
+	// UpdateIdentities() function.
+	idCache cache.IdentityCache
+
+	// map key is the string representation of the selector being cached.
+	selectors map[string]IdentitySelector
+}
+
+// NewSelectorCache creates a new SelectorCache.
+func NewSelectorCache() *SelectorCache {
+	return &SelectorCache{
+		idCache:   cache.GetIdentityCache(),
+		selectors: make(map[string]IdentitySelector),
+	}
+}
+
+var (
+	// selectorCache is the global selector cache. Additional ones
+	// are only created for testing.
+	selectorCache  = NewSelectorCache()
+	emptySelection = make([]identity.NumericIdentity, 0)
```

Comment:
Don't know how often this will be used, but is more cache efficient if we don't need multiple copies of empty slices (both slice header and the backing memory). We could maybe expand this to only keep one copy of any kind of a slice. We already keep them sorted, so it might be possible to build an index to them. I added this comment and changed the definition slightly as a nil slice should work as well:
```
	// Empty slice of numeric identities used for all selectors that select nothing
	emptySelection []identity.NumericIdentity
```